File: firmware/src/HotspotDetector.cpp
#include "HotspotDetector.h"

#include <algorithm>
#include <cstring>

namespace {

float computeMedian(const float* frame, int n) {
  static float sorted[kMaxFramePixels];
  std::memcpy(sorted, frame, sizeof(float) * n);
  std::sort(sorted, sorted + n);
  return sorted[n / 2];
}

}  // namespace

int HotspotDetector::detect(const float* frame, int rows, int cols, Hotspot* out,
                             int max_hotspots) const {
  const int n = rows * cols;
  const float ambient = computeMedian(frame, n);
  const float threshold = ambient + delta_c_;

  static bool hot_mask[kMaxFramePixels];
  static bool visited[kMaxFramePixels];
  for (int i = 0; i < n; ++i) {
    hot_mask[i] = frame[i] > threshold;
    visited[i] = false;
  }

  static int queue[kMaxFramePixels];
  int found = 0;

  for (int start = 0; start < n; ++start) {
    if (!hot_mask[start] || visited[start]) continue;

    // 4-connected iterative flood fill via explicit BFS queue (no recursion).
    int head = 0, tail = 0;
    queue[tail++] = start;
    visited[start] = true;

    int pixel_count = 0;
    double weighted_row_sum = 0.0;
    double weighted_col_sum = 0.0;
    double weight_sum = 0.0;
    float peak_temp = frame[start];

    while (head < tail) {
      const int idx = queue[head++];
      const int r = idx / cols;
      const int c = idx % cols;
      const float temp = frame[idx];

      ++pixel_count;
      weighted_row_sum += static_cast<double>(r) * temp;
      weighted_col_sum += static_cast<double>(c) * temp;
      weight_sum += temp;
      if (temp > peak_temp) peak_temp = temp;

      static const int dr[4] = {-1, 1, 0, 0};
      static const int dc[4] = {0, 0, -1, 1};
      for (int d = 0; d < 4; ++d) {
        const int nr = r + dr[d];
        const int nc = c + dc[d];
        if (nr < 0 || nr >= rows || nc < 0 || nc >= cols) continue;
        const int nidx = nr * cols + nc;
        if (hot_mask[nidx] && !visited[nidx]) {
          visited[nidx] = true;
          queue[tail++] = nidx;
        }
      }
    }

    if (pixel_count >= min_pixels_ && found < max_hotspots) {
      out[found].row = static_cast<float>(weighted_row_sum / weight_sum);
      out[found].col = static_cast<float>(weighted_col_sum / weight_sum);
      out[found].peak_temp_c = peak_temp;
      out[found].pixel_count = pixel_count;
      ++found;
    }
  }

  return found;
}

File: simulation/hotspot_detector.py
"""Hotspot detection algorithm (PROJECT_PLAN.md section 4). Python reference
implementation; firmware/src/HotspotDetector.cpp is a C++ port of this exact
algorithm.

    1. ambient = median(frame)
    2. hot_mask[i] = frame[i] > ambient + delta_c
    3. 4-connected iterative flood fill (BFS, explicit queue) into blobs
    4. discard blobs smaller than min_pixels
    5. report each blob's temperature-weighted centroid + peak temp
"""

from collections import deque

import numpy as np


def detect_hotspots(frame, delta_c, min_pixels):
    """frame: 2D numpy array (rows, cols) of temperatures in deg C.

    Returns a list of {row, col, peak_temp_c, pixel_count} dicts.
    """
    rows, cols = frame.shape
    ambient = np.median(frame)
    threshold = ambient + delta_c

    hot_mask = frame > threshold
    visited = np.zeros_like(hot_mask, dtype=bool)
    hotspots = []

    for r0 in range(rows):
        for c0 in range(cols):
            if not hot_mask[r0, c0] or visited[r0, c0]:
                continue

            queue = deque([(r0, c0)])
            visited[r0, c0] = True

            pixel_count = 0
            weighted_row_sum = 0.0
            weighted_col_sum = 0.0
            weight_sum = 0.0
            peak_temp = frame[r0, c0]

            while queue:
                r, c = queue.popleft()
                temp = frame[r, c]

                pixel_count += 1
                weighted_row_sum += r * temp
                weighted_col_sum += c * temp
                weight_sum += temp
                peak_temp = max(peak_temp, temp)

                for dr, dc in ((-1, 0), (1, 0), (0, -1), (0, 1)):
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < rows and 0 <= nc < cols and hot_mask[nr, nc] and not visited[nr, nc]:
                        visited[nr, nc] = True
                        queue.append((nr, nc))

            if pixel_count >= min_pixels:
                hotspots.append({
                    "row": weighted_row_sum / weight_sum,
                    "col": weighted_col_sum / weight_sum,
                    "peak_temp_c": float(peak_temp),
                    "pixel_count": pixel_count,
                })

    return hotspots
